File: server/reconcile.mjs
/**
 * Reconciliation: compare our synthetic Land-Twin parcels against the real
 * OpenStreetMap reference (buildings / land-use / roads).
 *
 * Produces, per parcel: whether a real building footprint sits on it, the
 * OSM building area vs our built-up estimate, whether OSM's land-use tag
 * agrees with ours, and the true distance to the nearest mapped road.
 * Plus dataset-level agreement rates.
 *
 * This measures agreement with community data — it is not a validation
 * against an official record (no such API exists for Indian parcels).
 */

const DEG2RAD = Math.PI / 180;

function ringOf(geom) {
  if (geom.type === "Polygon") return geom.coordinates[0];
  if (geom.type === "MultiPolygon") return geom.coordinates[0][0];
  return [];
}
function bbox(ring) {
  let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
  for (const [x, y] of ring) {
    minX = Math.min(minX, x); minY = Math.min(minY, y);
    maxX = Math.max(maxX, x); maxY = Math.max(maxY, y);
  }
  return [minX, minY, maxX, maxY];
}
function centroid(ring) {
  let x = 0, y = 0;
  for (const p of ring) { x += p[0]; y += p[1]; }
  return [x / ring.length, y / ring.length];
}
function pointInRing([px, py], ring) {
  let inside = false;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i++) {
    const [xi, yi] = ring[i], [xj, yj] = ring[j];
    if ((yi > py) !== (yj > py) && px < ((xj - xi) * (py - yi)) / (yj - yi) + xi) inside = !inside;
  }
  return inside;
}
/** planar area in m² (local equirectangular projection at the ring's latitude) */
function areaM2(ring) {
  if (ring.length < 4) return 0;
  const lat0 = centroid(ring)[1] * DEG2RAD;
  const kx = 111320 * Math.cos(lat0), ky = 110540;
  let a = 0;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i++) {
    const xi = ring[i][0] * kx, yi = ring[i][1] * ky;
    const xj = ring[j][0] * kx, yj = ring[j][1] * ky;
    a += xj * yi - xi * yj;
  }
  return Math.abs(a) / 2;
}
function distPointSegM([px, py], [ax, ay], [bx, by], kx, ky) {
  const Ax = ax * kx, Ay = ay * ky, Bx = bx * kx, By = by * ky, Px = px * kx, Py = py * ky;
  const dx = Bx - Ax, dy = By - Ay;
  const len2 = dx * dx + dy * dy || 1;
  let t = ((Px - Ax) * dx + (Py - Ay) * dy) / len2;
  t = Math.max(0, Math.min(1, t));
  const cx = Ax + t * dx, cy = Ay + t * dy;
  return Math.hypot(Px - cx, Py - cy);
}

const OSM_LANDUSE_MAP = {
  residential: "Residential",
  retail: "Commercial",
  commercial: "Commercial",
  industrial: "Industrial",
  education: "Institutional",
  religious: "Institutional",
  institutional: "Institutional",
  civic_admin: "Public/Semi-Public",
  railway: "Public/Semi-Public",
  recreation_ground: "Open Space",
  grass: "Open Space",
  park: "Open Space",
  forest: "Open Space",
  meadow: "Open Space",
  greenfield: "Vacant",
  brownfield: "Vacant",
  construction: "Vacant",
};

function median(xs) {
  if (!xs.length) return null;
  const s = [...xs].sort((a, b) => a - b);
  const m = s.length >> 1;
  return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2;
}

export function reconcile(parcels, reference) {
  const ref = reference?.features ?? [];
  const buildings = ref
    .filter((f) => f.properties.osm_kind === "building")
    .map((f) => { const r = ringOf(f.geometry); return { r, c: centroid(r), bb: bbox(r), area: areaM2(r), levels: f.properties.levels }; });
  const landuses = ref
    .filter((f) => f.properties.osm_kind === "landuse")
    .map((f) => { const r = ringOf(f.geometry); return { r, bb: bbox(r), tag: f.properties.landuse }; });
  const roads = ref
    .filter((f) => f.properties.osm_kind === "highway")
    .map((f) => ({ line: f.geometry.coordinates, name: f.properties.name, cls: f.properties.highway }));

  const perParcel = {};
  let withBuilding = 0, luComparable = 0, luAgree = 0;
  const builtDeltas = [];
  const roadDists = [];

  for (const pf of parcels.features) {
    const id = pf.properties.parcel_id;
    const r = ringOf(pf.geometry);
    const c = centroid(r);
    const pbb = bbox(r);
    const lat0 = c[1] * DEG2RAD;
    const kx = 111320 * Math.cos(lat0), ky = 110540;

    // buildings whose centroid falls inside this parcel
    const inside = buildings.filter(
      (b) => b.c[0] >= pbb[0] && b.c[0] <= pbb[2] && b.c[1] >= pbb[1] && b.c[1] <= pbb[3] && pointInRing(b.c, r),
    );
    const osmBuildingAreaSqm = inside.reduce((s, b) => s + b.area, 0);
    const osmBuildingPresent = inside.length > 0;
    if (osmBuildingPresent) withBuilding++;

    const ourBuiltUp = typeof pf.properties.built_up_area_sqm === "number" ? pf.properties.built_up_area_sqm : null;
    const builtUpDelta = osmBuildingPresent && ourBuiltUp != null ? Math.round((ourBuiltUp - osmBuildingAreaSqm) * 10) / 10 : null;
    if (builtUpDelta != null) builtDeltas.push(Math.abs(builtUpDelta));

    // OSM land-use polygon containing the parcel centroid
    let osmLanduse = null;
    for (const lu of landuses) {
      if (c[0] >= lu.bb[0] && c[0] <= lu.bb[2] && c[1] >= lu.bb[1] && c[1] <= lu.bb[3] && pointInRing(c, lu.r)) {
        osmLanduse = lu.tag; break;
      }
    }
    let landuseMatch = null;
    if (osmLanduse) {
      const mapped = OSM_LANDUSE_MAP[osmLanduse] ?? null;
      if (mapped) {
        luComparable++;
        landuseMatch = mapped === pf.properties.land_use;
        if (landuseMatch) luAgree++;
      }
    }

    // nearest road (metres)
    let nearestRoadM = null, nearestRoadName = null, nearestRoadCls = null;
    for (const rd of roads) {
      for (let i = 1; i < rd.line.length; i++) {
        const d = distPointSegM(c, rd.line[i - 1], rd.line[i], kx, ky);
        if (nearestRoadM == null || d < nearestRoadM) { nearestRoadM = d; nearestRoadName = rd.name; nearestRoadCls = rd.cls; }
      }
    }
    if (nearestRoadM != null) roadDists.push(nearestRoadM);

    perParcel[id] = {
      osm_building_present: osmBuildingPresent,
      osm_building_count: inside.length,
      osm_building_area_sqm: Math.round(osmBuildingAreaSqm),
      our_built_up_area_sqm: ourBuiltUp,
      built_up_delta_sqm: builtUpDelta,
      osm_landuse: osmLanduse,
      our_land_use: pf.properties.land_use ?? null,
      landuse_match: landuseMatch,
      nearest_road_m: nearestRoadM == null ? null : Math.round(nearestRoadM * 10) / 10,
      nearest_road_name: nearestRoadName ?? null,
      nearest_road_class: nearestRoadCls ?? null,
    };
  }

  const total = parcels.features.length;
  return {
    generated_at: new Date().toISOString(),
    reference_source: reference?.meta?.source ?? "OpenStreetMap",
    reference_loaded_at: reference?.meta?.fetched_at ?? null,
    summary: {
      parcels_total: total,
      osm_buildings_in_area: buildings.length,
      osm_landuse_polygons: landuses.length,
      osm_roads: roads.length,
      parcels_with_osm_building: withBuilding,
      building_presence_rate: total ? Math.round((withBuilding / total) * 1000) / 10 : 0,
      landuse_comparable_parcels: luComparable,
      landuse_agreements: luAgree,
      landuse_agreement_rate: luComparable ? Math.round((luAgree / luComparable) * 1000) / 10 : null,
      median_abs_builtup_delta_sqm: median(builtDeltas),
      median_nearest_road_m: median(roadDists) == null ? null : Math.round(median(roadDists) * 10) / 10,
    },
    note:
      "Agreement with OpenStreetMap community data. Our parcel boundaries are a derived grid, " +
      "so building/land-use overlap is indicative. This is NOT validation against an official land record.",
    perParcel,
  };
}

File: backend/app/reconcile.py
"""Reconciliation: compare the demo (synthetic) parcels against the real OSM reference.

Per parcel: is a real building footprint on it, OSM building area vs our
built-up estimate, does OSM's land-use tag agree with ours, and the true
distance to the nearest mapped road — plus dataset-level agreement rates.
This measures agreement with community data; it is NOT validation against an
official record (no such API exists for Indian parcels).
"""

from datetime import datetime, timezone

from .geo import (
    area_m2,
    bbox_of_ring,
    centroid_of,
    dist_point_seg_m,
    js_round,
    point_in_ring,
    ring_of,
    round_to,
    scale_at,
)

OSM_LANDUSE_MAP = {
    "residential": "Residential", "retail": "Commercial", "commercial": "Commercial",
    "industrial": "Industrial", "education": "Institutional", "religious": "Institutional",
    "institutional": "Institutional", "civic_admin": "Public/Semi-Public", "railway": "Public/Semi-Public",
    "recreation_ground": "Open Space", "grass": "Open Space", "park": "Open Space",
    "forest": "Open Space", "meadow": "Open Space", "greenfield": "Vacant",
    "brownfield": "Vacant", "construction": "Vacant",
}


def _median(xs: list[float]) -> float | None:
    if not xs:
        return None
    s = sorted(xs)
    m = len(s) // 2
    return s[m] if len(s) % 2 else (s[m - 1] + s[m]) / 2


def reconcile(parcels: dict, reference: dict | None) -> dict:
    ref = (reference or {}).get("features", [])
    buildings = []
    for f in ref:
        if f["properties"].get("osm_kind") == "building":
            r = ring_of(f["geometry"])
            if r:
                buildings.append({"r": r, "c": centroid_of(r), "area": area_m2(r)})
    landuses = []
    for f in ref:
        if f["properties"].get("osm_kind") == "landuse":
            r = ring_of(f["geometry"])
            if r:
                landuses.append((r, bbox_of_ring(r), f["properties"].get("landuse")))
    roads = [
        (f["geometry"]["coordinates"], f["properties"].get("name"), f["properties"].get("highway"))
        for f in ref
        if f["properties"].get("osm_kind") == "highway"
    ]

    per_parcel: dict[str, dict] = {}
    with_building = lu_comparable = lu_agree = 0
    built_deltas: list[float] = []
    road_dists: list[float] = []

    for pf in parcels["features"]:
        pid = pf["properties"]["parcel_id"]
        r = ring_of(pf["geometry"])
        c = centroid_of(r)
        pbb = bbox_of_ring(r)
        kx, ky = scale_at(c[1])

        inside = [
            b for b in buildings
            if pbb[0] <= b["c"][0] <= pbb[2] and pbb[1] <= b["c"][1] <= pbb[3] and point_in_ring(b["c"], r)
        ]
        osm_area = sum(b["area"] for b in inside)
        present = len(inside) > 0
        if present:
            with_building += 1

        our = pf["properties"].get("built_up_area_sqm")
        our = our if isinstance(our, (int, float)) else None
        delta = round_to(our - osm_area, 1) if present and our is not None else None
        if delta is not None:
            built_deltas.append(abs(delta))

        osm_landuse = None
        for lr, bb, tag in landuses:
            if bb[0] <= c[0] <= bb[2] and bb[1] <= c[1] <= bb[3] and point_in_ring(c, lr):
                osm_landuse = tag
                break
        match = None
        if osm_landuse:
            mapped = OSM_LANDUSE_MAP.get(osm_landuse)
            if mapped:
                lu_comparable += 1
                match = mapped == pf["properties"].get("land_use")
                if match:
                    lu_agree += 1

        road_m = road_name = road_cls = None
        for line, name, cls in roads:
            for i in range(1, len(line)):
                d = dist_point_seg_m(c, line[i - 1], line[i], kx, ky)
                if road_m is None or d < road_m:
                    road_m, road_name, road_cls = d, name, cls
        if road_m is not None:
            road_dists.append(road_m)

        per_parcel[pid] = {
            "osm_building_present": present,
            "osm_building_count": len(inside),
            "osm_building_area_sqm": js_round(osm_area),
            "our_built_up_area_sqm": our,
            "built_up_delta_sqm": delta,
            "osm_landuse": osm_landuse,
            "our_land_use": pf["properties"].get("land_use"),
            "landuse_match": match,
            "nearest_road_m": None if road_m is None else round_to(road_m, 1),
            "nearest_road_name": road_name,
            "nearest_road_class": road_cls,
        }

    total = len(parcels["features"])
    med_road = _median(road_dists)
    return {
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "reference_source": ((reference or {}).get("meta") or {}).get("source", "OpenStreetMap"),
        "reference_loaded_at": ((reference or {}).get("meta") or {}).get("fetched_at"),
        "summary": {
            "parcels_total": total,
            "osm_buildings_in_area": len(buildings),
            "osm_landuse_polygons": len(landuses),
            "osm_roads": len(roads),
            "parcels_with_osm_building": with_building,
            "building_presence_rate": round_to(with_building / total * 100, 1) if total else 0,
            "landuse_comparable_parcels": lu_comparable,
            "landuse_agreements": lu_agree,
            "landuse_agreement_rate": round_to(lu_agree / lu_comparable * 100, 1) if lu_comparable else None,
            "median_abs_builtup_delta_sqm": _median(built_deltas),
            "median_nearest_road_m": None if med_road is None else round_to(med_road, 1),
        },
        "note": (
            "Agreement with OpenStreetMap community data. Our parcel boundaries are a derived grid, "
            "so building/land-use overlap is indicative. This is NOT validation against an official land record."
        ),
        "perParcel": per_parcel,
    }
